File: store/738.单调递增的数字.c
/*
 * @lc app=leetcode.cn id=738 lang=c
 *
 * [738] 单调递增的数字
 */

// @lc code=start

int monotoneIncreasingDigits(int n)
{
  // 法一：暴力方法
  // int i, num, max, flag, res;
  // for (i = 0; i <= n; i++)
  // {
  //   num = i;
  //   max = num % 10;
  //   flag = 1;
  //   while (num)
  //   {
  //     if (max >= num % 10)
  //       max = num % 10;
  //     else
  //     {
  //       flag = 0;
  //       break;
  //     }
  //     num /= 10;
  //   }
  //   if (flag == 1)
  //     res = i;
  // }
  // return res;
  // 法二：贪心
  int nums[10], num, top, i, sum, flag;
  top = sum = 0;
  flag = -1;
  num = n;
  while (num)
  {
    nums[top++] = num % 10;
    num /= 10;
  }
  for (i = 1; i < top; i++)
  {
    if (nums[i] > nums[i - 1])
    {
      nums[i]--;
      flag = i - 1;
    }
  }
  for (i = 0; i <= flag; i++)
    nums[i] = 9;
  for (i = top - 1; i >= 0; i--)
  {
    if (nums[i] != 0)
    {
      sum *= 10;
      sum += nums[i];
    }
  }

  return sum;
}
// @lc code=end

File: store/738.单调递增的数字.js
/*
 * @lc app=leetcode.cn id=738 lang=javascript
 *
 * [738] 单调递增的数字
 */

// @lc code=start
/**
 * @param {number} n
 * @return {number}
 */
const isAdd = (num) => {
  let last, cur;
  // 另一种方法
  while (num) {
    cur = num % 10;
    num = Math.floor(num / 10);
    last = num % 10;
    if (last > cur) return false;
  }
  return true;
};
var monotoneIncreasingDigits = function (n) {
  // 法一：暴力
  // for (let i = n; i > 0; i--) {
  //   if (isAdd(i)) return i;
  // }
  // 法二：贪心
  let arr = String(n).split(""),
    flag = 0;
  for (let i = arr.length - 1; i > -1; i--) {
    if (Number(arr[i]) < Number(arr[i - 1])) {
      arr[i - 1] -= 1;
      flag = i;
    }
  }
  if (flag) for (let i = flag; i < arr.length; i++) arr[i] = "9";
  return Number(arr.join(""));
};
// @lc code=end
